### ISSUE TITLE
Concerns around error label and message

### ISSUE THREAD

**Issue by marcoscaceres** [gpuweb/gpuweb#2633]
Hi WebGPU WG,
I'm writing to raise several concerns about `GPUObjectBase`'s `label` and `GPUDeviceLostInfo`'s `message`. 

Tl;dr the concerns with these attributes are: 

 * Could serve as potential fingerprinting vectors. 
 * Generally not localized, and could end up being displayed to users - leading to a less than ideal user experience!
 * Non-idiomatic use of `undefined` for a WebIDL attribute value.
 * Includes errors/warnings that could just be presented in the developer console

Quick solution: To remedy the above, I strongly suggest removing these attributes from the specification. 

## Potential fingerprinting vector
Because "[message](https://www.w3.org/TR/webgpu/#dom-gpudevicelostinfo-message) set to an implementation-defined value", it risks being unique per implementation. This adds to the fingerprinting surface and variance between user agents. 

For `label`, the spec even notes:

> label which can be used by development tools (such as error/warning messages, browser developer tools, or platform debugging utilities) to identify the underlying [internal object](https://www.w3.org/TR/webgpu/#internal-object) to the developer. It has no specified format, and therefore cannot be reliably machine-parsed.

Then it shouldn't be part of the spec. The fact that it is there means that **it will most definitely will be parsed and it will be relied on by developers** leading to unexpected interoperability issues. 

## Localization
Generally, developer facing error messages are not localized. Consider, Geolocation suffers from this issue across all user agents. 

As happened with the Geolocation API, these messages end up being shown to end users. Developers often don't realize that these messages are not localized. For example in Blink (same in WebKit, btw... and also in Gecko):
https://chromium.googlesource.com/chromium/blink/+/refs/heads/main/Source/modules/geolocation/Geolocation.cpp#47

Sites are relying these, which means they can't be changed. Please avoid making this same mistake in WebGPU.

## Non-idiomatic use of undefined. 
Consider: nowhere else in the platform does one find `attribute undefined` as a type (or as a union), as having a property that is defined yet `undefined` is highly confusing.
 
I don't recommend the following either, but even Gecko has set `message` and `label` to nullable (`USVString?`) (also, the choice of `USVString` seems odd... should probably be a`DOMString`... but dropping it entirely is better):  
https://searchfox.org/mozilla-central/source/dom/webidl/WebGPU.webidl#51 

## Show errors/warnings directly in the developer console
Instead of exposing `GPUObjectBase`'s `label` and `GPUDeviceLostInfo`'s `message`, please recommend user agents route developer messages directly to the developer console. Presenting info/warning/error messages in the developer console without also exposing the a `.message` in the API is already common practice in the platform. 

For example: 
<img width="707" alt="Screen Shot 2022-03-02 at 3 16 27 pm" src="https://user-images.githubusercontent.com/870154/156294038-6a93ffdf-7552-4bf4-a77f-764c817f28cb.png">

The above is not from an API, the browser just does the right thing. There are tens, if not hundreds, of these little helpful warnings that show up in the developer console. They are easy to implement in user agents too and provide the best developer experience (including direct access to where the error occurred for debugging etc.).

**Comment by litherum** [comment]
`object.label` isn't supposed to hold values produced by the browser; it holds values produced by the website. The reason it exists is for a website to say something like "This is my command buffer where I produce 'screen space ambient occlusion' values, so I'll set the command buffer's label to the string "SSAO" and then when I attach to the browser with RenderDoc or whatever, I'll be able to find my SSAO command buffer among all the command buffers the browser creates."

**Comment by litherum** [comment]
> Presenting info/warning/error messages in the developer console without also exposing the a .message in the API is already common practice in the platform.

Is this what WebAssembly does? They're in a similar situation as us; the site provides a program to the browser, written in some language, and the browser has to try to make sense of the program.

**Comment by marcoscaceres** [comment]
Thanks @litherum for the clarifications - that was really helpful, and my apologies in advance as I'm very much a novice in this space. My reading from the spec was that these were user agent generated messages (particularly `message` seems to be set by the user agent in the "[lose a device](https://www.w3.org/TR/webgpu/#lose-the-device)" algorithm). It may be worth clarifying the above in the spec a little bit more.

Having said that, I'm still wondering if `USVString?` would be more appropriate as a type? I've asked more generally about this kind of usage in https://github.com/whatwg/webidl/issues/1112.

**Comment by litherum** [comment]
It looks like @kainino0x intentionally used undefined: https://github.com/gpuweb/gpuweb/commit/89b0dba025e3a6a550ce9f82de6e07f3c1a6b008 but there doesn’t seem to be any discussion about why
 It’s preferable in the commit message.

**Comment by marcoscaceres** [comment]
Thanks @litherum. I did spot that too. 

@kainino0x, if there is a solid rationale for it, it might just be worth adding a note (as I couldn't find any other IDL attribute in the platform that uses this union or returns `undefined` - that not to say it's wrong, just, um, "eyebrow raising":)). I was pinged independently by an implementer who was quite surprised/confused by it, which is what prompted me to file this issue in the first place 🙏. 

Thanks again for considering my input!

**Comment by kainino0x** [comment]
Sorry, will catch up on this thread soon.
As to the last question, yes, it's intentional because GPUObjectDescriptorBase takes an optional `USVString label;` and therefore can be read/written as `(USVString or undefined)`, so `GPUObjectBase.label` seems that it should match, rather than being instead `USVString?` (`USVString or null`).

**Comment by litherum** [comment]
I think there are 3 concerns in this thread, and we should probably split them out:

1. Reconsider `GPUObjectBase.label`. I think this has been resolved; there's no need to reconsider it
2. Reconsider `GPUCompilationMessage.message`, `GPUDeviceLostInfo.message`, and `GPUValidationError.message` because fingerprinting, localization, and what do we expect developers to actually do with these strings that they couldn't do if the strings were logged to the console instead
3. Reconsider `(USVString or undefined)` to possibly become `USVString?`

**Comment by kainino0x** [comment]
Thanks @litherum for responding on this thread; agree with your assessment.

On #2: AFAIK exception messages are not normally localized, e.g. in chrome (or safari):
![image](https://user-images.githubusercontent.com/606355/157511068-e0619dce-b13b-48e3-87cd-58a7a22f1dc9.png)
So for anything we don't think would be usefully surfaced to the user, it seems we should follow that, or not expose a message (fine **in some cases**<sup>1</sup>).
Anything that would be usefully surfaced to the user (GPUDeviceLostInfo.message especially) should probably be localized by the browser.
GPUCompilationMessage.message is unclear, as it's rarely exposed to the user, but would be exposed on ShaderToy-like sites. It probably makes sense to localize in an ideal world.

Fingerprinting is a non-concern to me. The contents of GPUCompilationMessage.message and GPUValidationError.message expose no identifying information except which browser engine you use, which is already clearly visible by the presence/absence of other APIs. GPUDeviceLostInfo.message is only seen in exceptional circumstances and at worst would only ever expose vague info like "GPUDevice destroyed", "graphics device unavailable", or "ran out of graphics resources" (and most can't be exposed except by some very exceptional external event).

<sup>1</sup> and to answer a question above: The [telemetry use case](https://github.com/gpuweb/gpuweb/blob/main/design/ErrorHandling.md#telemetry) still needs information to be exposed outside dev tools (at least in `uncapturederror`; popErrorScope doesn't really need it). Technically we could report details like that through the [Reporting API](https://developer.mozilla.org/en-US/docs/Web/API/Reporting_API), but that doesn't seem to help anything.

**Comment by Kangz** [comment]
From the split of the issue @litherum made, I think 1) and 2) are answered so we only need to consider 3) and it's small enough that we could split it to its own issue (though TBH I think we could also just say that things stay as is since it was already discussed previously to go from `USVString?` to `USVString or undefined`)

**Comment by kdashg** [comment]
["Note: Although dictionary types can in general be nullable, they cannot when used as the type of an operation argument or a dictionary member."](https://webidl.spec.whatwg.org/#idl-nullable-type)

**Comment by Kangz** [comment]
@kdashg, I'm not sure this quote answers point 3) since `USVString` isn't a dictionary.

**Comment by litherum** [comment]
- fingerprintinglocalization, and what do we expect developers to actually do with these strings that they couldn't do if the strings were logged to the console instead

**Comment by litherum** [comment]
- fingerprinting: I don't think there is any fingerprinting concern here.
- localization: The contents of these messages are not spec'd, so browsers are free to localize them if they desire.
- what do we expect developers to actually do with these strings that they couldn't do if the strings were logged to the console instead: This makes sense to me. For almost all use cases, even if the page showed these error strings to the user, the user is most likely not in a situation to act upon them (or possibly even understand them!). So I do think this is a valid argument, with the goal to reduce API surface area.

**Comment by mehmetoguzderin** [comment]
The localization aspect of this issue might soon relate to i18n feedback, where they suggest https://github.com/w3c/i18n-activity/issues/1503 exposing language and direction information. This suggestion comes from WebGPU being a Web spec where intended-as-human-readable strings can end up plugging into a localizable strings system where they end up with a corresponding localization, just like the UI of the spec. Also, see this https://github.com/gpuweb/gpuweb/issues/2546#issuecomment-1088138132 for another reflection on this suggestion. These fields sit within the message because a user speaking or having to speak multiple languages can change their language mid-browsing experience. And, I do not think there is any limitation on browsers not to decide to adjust their localization in that page according to that.

**Comment by mehmetoguzderin** [comment]
In addition, I think it is better to keep messages, not adopt inflexible error codes, and to be current with how things could better be as suggested by the original post and to be by i18n team, just have language and directionality fields to accompany any human-readable (which brings in the possibility that browser can decide to localize unless not explicitly stated because browsers do have fantastic localization facilities) string.

**Comment by Kangz** [comment]
@mehmetoguzderin while this is certainly linked to this issue, I think we first need to resolve this, and only then discuss internationalization topics. If the strings are not given to JS the internationalization question is solved (as far as the API is concerned).

> what do we expect developers to actually do with these strings that they couldn't do if the strings were logged to the console instead: This makes sense to me. For almost all use cases, even if the page showed these error strings to the user, the user is most likely not in a situation to act upon them (or possibly even understand them!). So I do think this is a valid argument, with the goal to reduce API surface area.

There are definite use cases for giving the strings to JS (as discussed in the meeting). Showing compilation errors in WGSL code editors are a big one. But for validation errors messages, the usecase that @toji discussed with CodePen looking at the JS exception message and showing it to the user works as well.

That's why I strongly think we should keep both the WGSL and validation error messages exposed at the API level (though the extent of my conviction didn't show in the meeting). `GPUDeviceLostInfo` I also think we should keep but have less strong of an opinion.

This issue should then only be editorial about optional `USVString` and the internationalization discussions should be in new issues.

**Comment by litherum** [comment]
Regarding error codes:

1. For WGSL compiler errors, we have existing evidence (Clang error messages) that innovation in the error message space is valuable, and indeed, important.
2. I'm pessimistic that the CG would even realistically be _able_ to create standardized error codes.

That being said, even if the spec doesn't require error codes, each implementation can implement and abide by their own error codes. I think MSVC does this. This kind of solution is probably sufficient to address the i18n desires.

**Comment by mehmetoguzderin** [comment]
@Kangz Agreed, I think if it is possible to render OPs localization section as "can you solve this by just removing?" binary answer question, then we can compartmentalize lang, dir topic to be discrete issue/PR once we have the full feedback on that.

**Comment by toji** [comment]
I've put up #2784 to resolve the concern about using `undefined` and clarified who supplies the label and how it might be used.

**Comment by Kangz** [comment]
Thanks, I think we can now close this issue as discussed in last meeting. I18n is in #2780, `USVString?` vs `undefined` is #2784 and the rest is N/A.
